packages/hooks/src/useCollectionQuery/useCollectionQuery.ts

```diff
@@ -0,0 +1,343 @@
+   */
+  subscription?: ListSubscription<TSubscription>;
+}
+
+interface ListSubscription<TSubscription> {
+  /**
+   * The graphQL subscription that listens for more data. This query should
+   * return a single Node that matches the data structure in
+   * `getCollectionByPath<TQuery>(...).edges.node` and
+   * `getCollectionByPath<TQuery>(...).nodes
+   */
+  document: DocumentNode;
+
+  /**
+   * A list of variables to pass into the apollo `subscribeToMore` function.
+   */
+  options?: Pick<SubscribeToMoreOptions<TSubscription>, "variables">;
+
+  /**
+   * A function that returns the location where the `Node` is located on the
+   * `TSubscription` object.
+   *
+   * It should return a single Node that matches the data structure in
+   * `getCollectionByPath<TQuery>(...).edges.node` and
+   * `getCollectionByPath<TQuery>(...).nodes
+   */
+  getNodeByPath: GetNodeByPath<TSubscription>;
+}
+
+interface Collection {
+  edges?: Edge[];
+  nodes?: Node[];
+  pageInfo: {
+    endCursor?: string | undefined;
+    hasNextPage: boolean;
+    [otherProperties: string]: unknown;
+  };
+  totalCount: number;
+  [otherProperties: string]: unknown;
+}
+
+interface CollectionQueryResult<TQuery> {
+  data: TQuery | undefined;
+  error: ApolloError | undefined;
+  loadingRefresh: boolean;
+  loadingNextPage: boolean;
+  loadingInitialContent: boolean;
+  refresh(): void;
+  nextPage(): void;
+}
+
+type GetCollectionByPathFunction<TQuery> = (
+  data: TQuery | undefined,
+) => Collection | undefined;
+
+type GetNodeByPath<TSubscription> = (
+  data: TSubscription | undefined,
+) => Node | undefined;
+
+export function useCollectionQuery<TQuery, TSubscription = undefined>({
+  query,
+  queryOptions,
+  getCollectionByPath,
+  subscription,
+}: UseCollectionQueryArguments<
+  TQuery,
+  TSubscription
+>): CollectionQueryResult<TQuery> {
+  const {
+    data,
+    loading,
+    refetch,
+    error,
+    fetchMore,
+    subscribeToMore,
+  } = useQuery<TQuery>(query, queryOptions);
+
+  const isMounted = useIsMounted();
+  const [loadingRefresh, setLoadingRefresh] = useState<boolean>(false);
+  const [loadingNextPage, setLoadingNextPage] = useState<boolean>(false);
+  const loadingInitialContent = loading && !loadingRefresh && !loadingNextPage;
+
+  const refresh = useCallback(() => {
+    if (loadingInitialContent || loadingRefresh) {
+      return;
+    }
+
+    setLoadingRefresh(true);
+    refetch()
+      .catch(err => Configuration.errorNotifier("Refetch Error", err))
+      .finally(() => {
+        if (isMounted.current) {
+          setLoadingRefresh(false);
+        }
+      });
+  }, [
+    loadingInitialContent,
+    loadingRefresh,
+    setLoadingRefresh,
+    refetch,
+    isMounted,
+  ]);
+
+  const nextPage = useCallback(() => {
+    if (loadingInitialContent || loadingRefresh || loadingNextPage) {
+      return;
+    }
+
+    const pageInfo = getCollectionByPath(data)?.pageInfo;
+
+    if (!pageInfo || !pageInfo.hasNextPage) {
```

Comment:
Would `!pageInfo?.hasNextPage` work here?

Reply:
@seanhealy: 
It certainly could and since you're maintianing this repo if you feel strongly I'll change it, but I think the above form is more readable and easier to reason around.

The reason is that `pageInfo` not existing and `hasNextPage` are two different states that *could* have two  different code paths. They happen to share a code path right now but that doesn't have to be the case.

`!pageInfo?.hasNextPage` leaves it up to the reader to decide if the auther of the code explicitly chose that `!pageInfo` should return `nil` or if it was an accident because `pageInfo` was optionally `null` and to get `hasNextPage` to work the auther just added `?`

Reply:
Thoughts?